src/lib.rs

```diff
@@ -222,15 +222,13 @@ impl AsyncRead for Serial {
         cx: &mut Context<'_>,
         buf: &mut ReadBuf<'_>,
     ) -> Poll<io::Result<()>> {
-        loop {
-            let mut guard = ready!(self.io.poll_read_ready(cx))?;
-            match guard.try_io(|_| {
-                let read = self.io.get_ref().read(buf.initialize_unfilled())?;
-                return Ok(buf.advance(read));
-            }) {
-                Ok(result) => return Poll::Ready(result),
-                Err(_would_block) => continue,
-            }
+        let mut guard = ready!(self.io.poll_read_ready(cx))?;
+        match guard.try_io(|_| {
+            let read = self.io.get_ref().read(buf.initialize_unfilled())?;
+            return Ok(buf.advance(read));
+        }) {
+            Ok(result) => return Poll::Ready(result),
+            Err(_would_block) => return Poll::Pending,
```

Comment:
This breaks as the `read()` can return would block in which case you go pending without the async fd actually registered with the runtime;  The original version with the loop is correct as `read()` returns would_block then the `poll_read_ready()` call should return Pending and everything should work nicely :).

fwiw i hit this problem in practise while playing with this branch and a fdti usb dongle in linux

Reply:
I'm confused about this: if the read returns EWOULDBLOCK doesn't that imply a bug somewhere else? We can't arrive at line 227 without being signaled by tokio that there are in fact _some_ bytes to read. I haven't been using this branch but I have some code that is morally equiv that I have been testing with ftdi dongles (snap!) and I haven't seen this condition triggered (yet).

Reply:
It is possible to get a false positive from `poll_read_ready` and get the would block error, but the snippet is still incorrect — you have to loop around and call `poll_read_ready` again if that happens.

Reply:
I see what you are saying. Thanks for making that connection. I also completely misread the sample code for `AsyncFd`. I guess I'd just gotten lucky :)

Reply:
@berkowski Please look at PR: https://github.com/mnbbrown/tokio-serial/pull/1

Reply:
I can confirm I have been seeing this bug too and rfael's patch fixed it.